src/axom/primal/operators/intersect.hpp

```diff
@@ -614,6 +735,215 @@ AXOM_HOST_DEVICE bool intersect(const Plane<T, 3>& p,
+      t.push_back(tc[i]);
+      u.push_back(uc[i]);
+      v.push_back(vc[i]);
+    }
+  }
+
+  return !t.empty();
+}
+
+template <typename T>
+AXOM_HOST_DEVICE bool intersect(const Ray<T, 3>& ray,
+                                const NURBSPatch<T, 3>& patch,
+                                axom::Array<T>& t,
+                                axom::Array<T>& u,
+                                axom::Array<T>& v,
+                                double tol = 1e-8,
+                                double EPS = 1e-8,
+                                bool isHalfOpen = false)
+{
+  // Check a bounding box of the entire NURBS first
+  Point<T, 3> ip;
+  if(!intersect(ray, patch.boundingBox(), ip))
+  {
+    return false;
+  }
+
+  // Decompose the NURBS patch into Bezier patches
+  auto beziers = patch.extractBezier();
+
+  axom::Array<T> knot_vals_u = patch.getKnots_u().getUniqueKnots();
+  axom::Array<T> knot_vals_v = patch.getKnots_v().getUniqueKnots();
+
+  const auto num_knot_span_u = knot_vals_u.size() - 1;
+  const auto num_knot_span_v = knot_vals_v.size() - 1;
+
+  // Store candidate intersections
+  axom::Array<T> tc, uc, vc;
+
+  // Check each Bezier patch, and scale the intersection parameters
+  //  back into the span of the original NURBS patch
+  for(int i = 0; i < num_knot_span_u; ++i)
+  {
+    for(int j = 0; j < num_knot_span_v; ++j)
+    {
+      auto& bezier = beziers[i * num_knot_span_v + j];
+
+      // Store candidate intersections from each Bezier patch
+      axom::Array<T> tcc, ucc, vcc;
+      intersect(ray, bezier, tcc, ucc, vcc, tol, EPS);
+
+      // Scale the intersection parameters back into the span of the NURBS patch
+      for(int k = 0; k < tcc.size(); ++k)
+      {
+        tc.push_back(tcc[k]);
+        uc.push_back(knot_vals_u[i] +
+                     ucc[k] * (knot_vals_u[i + 1] - knot_vals_u[i]));
+        vc.push_back(knot_vals_v[j] +
+                     vcc[k] * (knot_vals_v[j + 1] - knot_vals_v[j]));
+      }
+    }
+  }
+
+  // Do a second pass to remove duplicates from uc, vc
+  const double sq_EPS = EPS * EPS;
+
+  // The number of reported intersection points will be small,
+  //  so we don't need to fully sort the list
+
+  double max_u_knot = patch.getKnots_u()[patch.getKnots_u().getNumKnots() - 1];
+  double max_v_knot = patch.getKnots_v()[patch.getKnots_v().getNumKnots() - 1];
+
+  for(int i = 0; i < tc.size(); ++i)
+  {
+    // Also remove any intersections on the half-interval boundaries
+    if(isHalfOpen && (uc[i] >= max_u_knot - EPS || vc[i] >= max_v_knot - EPS))
+    {
+      continue;
+    }
+
+    Point<T, 2> uv({uc[i], vc[i]});
+
+    bool foundDuplicate = false;
+    for(int j = i + 1; !foundDuplicate && j < tc.size(); ++j)
+    {
+      if(squared_distance(uv, Point<T, 2>({uc[j], vc[j]})) < sq_EPS)
+      {
+        foundDuplicate = true;
+      }
+    }
+
+    if(!foundDuplicate)
+    {
+      t.push_back(tc[i]);
+      u.push_back(uc[i]);
+      v.push_back(vc[i]);
+    }
+  }
```

Comment:
I'm not sure if this particular code would be useful outside of this context, since it isn't just removing duplicate pairs of values, it's also checking them against the half-open intervals. We'd either have to iterate through the list twice, or add this logic as a utility function to `primal::NURBSPatch` specifically, which feels like overkill for something only used twice.